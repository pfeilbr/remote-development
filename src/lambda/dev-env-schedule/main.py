import boto3
import os
import json

region = os.environ.get("REGION")
instance_ids = json.loads(os.environ.get("INSTANCE_IDS"))
ec2 = boto3.client("ec2", region_name=region)
sns = boto3.client("sns", region_name=region)


def is_instance_running(instance_id):
    response = ec2.describe_instance_status(InstanceIds=[instance_id])
    try:
        return response["InstanceStatuses"][0]["InstanceState"]["Name"] == "running"
    except:
        return False


def instance_operation(event, context):
    # operation = getattr(event, 'operation', 'stop_instances')
    for instance_id in instance_ids:
        if is_instance_running(instance_id):
            ec2.stop_instances(InstanceIds=[instance_id])
            print("{} instance (instance_id={})".format("stop_instances", instance_id))
            notify(event, context)


def notify(event, context):
    topic_arn = region = os.environ.get("NOTIFY_TOPIC_ARN")
    sns.publish(
        TopicArn=topic_arn,
        Subject="stopped ec2 instance(s) ({})".format(str(instance_ids)),
        Message="stopped ec2 instance(s) with instance id(s) of {}".format(
            str(instance_ids)
        ),
    )
