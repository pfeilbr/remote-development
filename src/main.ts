
import * as events from '@aws-cdk/aws-events';
import * as targets from '@aws-cdk/aws-events-targets';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { PythonFunction } from '@aws-cdk/aws-lambda-python';
import * as sns from '@aws-cdk/aws-sns';
import * as subscriptions from '@aws-cdk/aws-sns-subscriptions';
import { App, Construct, Stack, StackProps, Duration, CfnOutput } from '@aws-cdk/core';

export class RemoteDevelopmentStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const notifyTopic = new sns.Topic(this, 'NotifyTopic');
    notifyTopic.addSubscription(new subscriptions.EmailSubscription('brian.pfeil@gmail.com'));


    const lambdaFn = new PythonFunction(this, 'ScheduledDevelopmentEnvironment', {
      entry: 'src/lambda/dev-env-schedule', // required
      index: 'main.py', // optional, defaults to 'index.py'
      handler: 'instance_operation', // optional, defaults to 'handler'
      timeout: Duration.seconds(300),
      runtime: lambda.Runtime.PYTHON_3_8, // optional, defaults to lambda.Runtime.PYTHON_3_7
      environment: {
        REGION: 'us-east-1',
        INSTANCE_IDS: '["i-06c49f207e012e481"]',
        NOTIFY_TOPIC_ARN: notifyTopic.topicArn,
      },
    });

    notifyTopic.grantPublish(lambdaFn);

    lambdaFn.addToRolePolicy(new PolicyStatement({
      resources: ['*'],
      actions: ['ec2:StartInstances', 'ec2:StopInstances', 'ec2:Describe*', 'ec2:List*', 'ec2:Get*'],
    }));

    // Run every day at 6PM UTC
    // See https://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html
    const rule = new events.Rule(this, 'Rule', {
      schedule: events.Schedule.expression('cron(0 4 * * ? *)'), // midnight daily
    });

    rule.addTarget(new targets.LambdaFunction(lambdaFn));

    new CfnOutput(this, 'InstanceOperationLambdaFunctionName', { value: lambdaFn.functionName });
    new CfnOutput(this, 'InstanceOperationLambdaFunctionArn', { value: lambdaFn.functionArn });
    new CfnOutput(this, 'CronRuleName', { value: rule.ruleName });
    new CfnOutput(this, 'CronRuleArn', { value: rule.ruleArn });
    new CfnOutput(this, 'NotifyTopicName', { value: notifyTopic.topicName });
    new CfnOutput(this, 'NotifyTopicArn', { value: notifyTopic.topicArn });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new RemoteDevelopmentStack(app, 'remote-development-stack', { env: devEnv });
// new RemoteDevelopmentStack(app, 'my-stack-prod', { env: prodEnv });

app.synth();