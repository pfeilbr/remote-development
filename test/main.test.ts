import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { RemoteDevelopmentConstruct } from '../src/main';

test('Snapshot', () => {
  const env = {
    account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  };

  const app = new App();
  const stack = new Stack(app, 'remote-development-stack', { env });
  new RemoteDevelopmentConstruct(stack, 'remote-development-construct');

  expect(stack).not.toHaveResource('AWS::S3::Bucket');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});