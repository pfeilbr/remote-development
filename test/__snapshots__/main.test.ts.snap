// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "CronRuleArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "Rule4C995B7F",
          "Arn",
        ],
      },
    },
    "CronRuleName": Object {
      "Value": Object {
        "Ref": "Rule4C995B7F",
      },
    },
    "InstanceOperationLambdaFunctionArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "ScheduledDevelopmentEnvironment82C14BD1",
          "Arn",
        ],
      },
    },
    "InstanceOperationLambdaFunctionName": Object {
      "Value": Object {
        "Ref": "ScheduledDevelopmentEnvironment82C14BD1",
      },
    },
    "NotifyTopicArn": Object {
      "Value": Object {
        "Ref": "NotifyTopic5FFA0016",
      },
    },
    "NotifyTopicName": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "NotifyTopic5FFA0016",
          "TopicName",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596ArtifactHash5ED491A9": Object {
      "Description": "Artifact hash for asset \\"f122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596\\"",
      "Type": "String",
    },
    "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596S3Bucket3E1DCE6A": Object {
      "Description": "S3 bucket for asset \\"f122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596\\"",
      "Type": "String",
    },
    "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596S3VersionKey7A90C284": Object {
      "Description": "S3 key for asset version \\"f122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "NotifyTopic5FFA0016": Object {
      "Type": "AWS::SNS::Topic",
    },
    "NotifyTopicbrianpfeilgmailcom92A1199C": Object {
      "Properties": Object {
        "Endpoint": "brian.pfeil@gmail.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "NotifyTopic5FFA0016",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Rule4C995B7F": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0 4 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "ScheduledDevelopmentEnvironment82C14BD1",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "RuleAllowEventRuletestScheduledDevelopmentEnvironment9EB6D177C6E90BC9": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ScheduledDevelopmentEnvironment82C14BD1",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "Rule4C995B7F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduledDevelopmentEnvironment82C14BD1": Object {
      "DependsOn": Array [
        "ScheduledDevelopmentEnvironmentServiceRoleDefaultPolicyBB6B7A7A",
        "ScheduledDevelopmentEnvironmentServiceRole2245AF2F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596S3Bucket3E1DCE6A",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596S3VersionKey7A90C284",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf122ab87bb0b795ae0b24dc00dd3d09239483722c721e243aa147b5bdbd18596S3VersionKey7A90C284",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "INSTANCE_IDS": "[\\"i-06c49f207e012e481\\"]",
            "NOTIFY_TOPIC_ARN": Object {
              "Ref": "NotifyTopic5FFA0016",
            },
            "REGION": "us-east-1",
          },
        },
        "Handler": "main.instance_operation",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ScheduledDevelopmentEnvironmentServiceRole2245AF2F",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduledDevelopmentEnvironmentServiceRole2245AF2F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledDevelopmentEnvironmentServiceRoleDefaultPolicyBB6B7A7A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "NotifyTopic5FFA0016",
              },
            },
            Object {
              "Action": Array [
                "ec2:StartInstances",
                "ec2:StopInstances",
                "ec2:Describe*",
                "ec2:List*",
                "ec2:Get*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledDevelopmentEnvironmentServiceRoleDefaultPolicyBB6B7A7A",
        "Roles": Array [
          Object {
            "Ref": "ScheduledDevelopmentEnvironmentServiceRole2245AF2F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
